pkg/operator/oidcdiscoveryendpoint/controller.go

```diff
@@ -0,0 +1,413 @@
+	}
+
+	auth.Spec.ServiceAccountIssuer = serviceAccountIssuer
+	_, err = r.configclientset.ConfigV1().Authentications().Update(context.TODO(), auth, metav1.UpdateOptions{})
+	return false, err
+}
+
+func (r *s3EndpointReconciler) getIssuerURL() string {
+	return fmt.Sprintf("https://s3-%s.amazonaws.com/%s", r.region, r.getBucketName())
+}
+
+func (r *s3EndpointReconciler) getBucketName() string {
+	return fmt.Sprintf("%s-oidc", r.infrastructureName)
+}
+
+func (r *s3EndpointReconciler) reconcileS3Endpoint() error {
+	r.logger.Debugf("reconciling S3 endpoint resources")
+
+	// Get the root secret and create an AWS client
+	secret, err := r.kubeclientset.CoreV1().Secrets(constants.CloudCredSecretNamespace).Get(context.TODO(), awsannotator.AWSCloudCredSecretName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	accessKey, ok := secret.Data[awsannotator.AwsAccessKeyName]
+	if !ok {
+		return fmt.Errorf("couldn't fetch key containing AWS_ACCESS_KEY_ID from cloud cred secret")
+	}
+
+	secretKey, ok := secret.Data[awsannotator.AwsSecretAccessKeyName]
+	if !ok {
+		return fmt.Errorf("couldn't fetch key containing AWS_ACCESS_KEY_ID from cloud cred secret")
+	}
+
+	awsClient, err := aws.NewClient(accessKey, secretKey, r.region, r.infrastructureName)
+	if err != nil {
+		return fmt.Errorf("error creating aws client: %v", err)
+	}
+
+	// Ensure bucket exists
+	bucketName := r.getBucketName()
+	createBucketInput := &s3.CreateBucketInput{
+		Bucket: awssdk.String(bucketName),
+	}
+
+	_, err = awsClient.CreateBucket(createBucketInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeBucketAlreadyExists:
+				r.logger.Errorf("bucket %s already exists but it not owned by us", bucketName)
+				return aerr
+			case s3.ErrCodeBucketAlreadyOwnedByYou:
+				r.logger.Debugf("bucket %s already exists and is owned by us", bucketName)
+			default:
+				return aerr
+			}
+		} else {
+			return err
+		}
+	} else {
+		r.logger.Debugf("bucket %s created", bucketName)
+	}
+
+	// Render and create the OIDC discovery JSON s3 object in the well-known location
+	issuerURL := r.getIssuerURL()
+	discoveryJSON := fmt.Sprintf(discoveryTemplate, issuerURL, issuerURL, keysURI)
+	discoveryPutObjInput := &s3.PutObjectInput{
+		ACL:    awssdk.String("public-read"),
+		Body:   awssdk.ReadSeekCloser(strings.NewReader(discoveryJSON)),
+		Bucket: awssdk.String(bucketName),
+		Key:    awssdk.String(discoveryURI),
+	}
+
+	_, err = awsClient.PutObject(discoveryPutObjInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			default:
+				return aerr
+			}
+		} else {
+			return err
+		}
+	}
+	r.logger.Debugf("discovery document updated")
+
+	// Extract the token signer and create keys JSON s3 object for jwks_uri
+	tokenSignerConfigMap, err := r.kubeclientset.CoreV1().ConfigMaps(tokenSignerConfigMapNamespace).Get(context.TODO(), tokenSignerConfigMapName, metav1.GetOptions{})
```

Comment:
Curious why you're using the kubeclientset as opposed to the controller-runtime Client? Has always seemed easier to work with to me.

Reply:
i had to create a clientset for the openshift config API (for accessing `Authentications`) so I just wanted to stay in that mindset.  no particular reason.  if you perfer the controller-runtime client, i switch to it for the kube API requests.

Reply:
as i attempted to switch to the controller client, i remember why i used the kubeclientset.  it was to get the `v1.EventInterface` needed for the `eventRecorder` on line 115